core/ldap_api.php

```diff
@@ -207,76 +189,107 @@ function ldap_escape_string( $p_string ) {
 }
 
 /**
- * Gets the value of a specific field from LDAP given the user name
- * and LDAP field name.
+ * Retrieves user data from LDAP and stores it in cache.
  *
- * @todo Implement caching by retrieving all needed information in one query.
- * @todo Implement logging to LDAP queries same way like DB queries.
+ * Uses a single LDAP query to retrieve the following fields:
+ * - email (mail)
+ * - realname {@see $g_ldap_realname_field}
  *
- * @param string $p_username The user name.
- * @param string $p_field    The LDAP field name.
- * @return string The field value or null if not found.
+ * @param string $p_username The username.
+ *
+ * @return array|false User data, false if not found or errors occurred
  */
-function ldap_get_field_from_username( $p_username, $p_field ) {
-	$t_ldap_organization    = config_get( 'ldap_organization' );
-	$t_ldap_root_dn         = config_get( 'ldap_root_dn' );
-	$t_ldap_uid_field		= config_get( 'ldap_uid_field' );
+function ldap_cache_user_data( $p_username ) {
+	global $g_cache_ldap_data;
 
-	$c_username = ldap_escape_string( $p_username );
+	# Returne cached data if available
+	if( isset( $g_cache_ldap_data[$p_username] ) ) {
+		return $g_cache_ldap_data[$p_username];
+	}
 
-	log_event( LOG_LDAP, 'Retrieving field \'' . $p_field . '\' for \'' . $p_username . '\'' );
+	log_event( LOG_LDAP, "Retrieving data for '$p_username' from LDAP server" );
 
 	# Bind
 	log_event( LOG_LDAP, 'Binding to LDAP server' );
 	$t_ds = @ldap_connect_bind();
 	if( $t_ds === false ) {
 		ldap_log_error( $t_ds );
-		return null;
+		return false;
 	}
 
 	# Search
-	$t_search_filter        = '(&' . $t_ldap_organization . '(' . $t_ldap_uid_field . '=' . $c_username . '))';
-	$t_search_attrs         = array( $t_ldap_uid_field, $p_field, 'dn' );
+	$t_ldap_organization = config_get( 'ldap_organization' );
+	$t_ldap_root_dn      = config_get( 'ldap_root_dn' );
+	$t_ldap_uid_field    = config_get( 'ldap_uid_field' );
+
+	$t_search_filter = '(&' . $t_ldap_organization
+		. '(' . $t_ldap_uid_field . '=' . ldap_escape_string( $p_username ) . '))';
+	$t_search_attrs = array(
+		'mail',
+		config_get( 'ldap_realname_field' )
+	);
 
 	log_event( LOG_LDAP, 'Searching for ' . $t_search_filter );
 	$t_sr = @ldap_search( $t_ds, $t_ldap_root_dn, $t_search_filter, $t_search_attrs );
 	if( $t_sr === false ) {
 		ldap_log_error( $t_ds );
 		ldap_unbind( $t_ds );
 		log_event( LOG_LDAP, 'ldap search failed' );
-		return null;
+		return false;
 	}
 
 	# Get results
-	$t_info = ldap_get_entries( $t_ds, $t_sr );
-	if( $t_info === false ) {
-		ldap_log_error( $t_ds );
-		log_event( LOG_LDAP, 'ldap_get_entries() returned false.' );
-		return null;
+	$t_entry = ldap_first_entry( $t_ds, $t_sr );
+	if( $t_entry === false ) {
+		log_event( LOG_LDAP, 'No matches found.' );
```

Comment:
It would be useful to include the search filter here as well, since logs tend to interleave across different requests.

Reply:
The search filter _is_ logged already, [just a few lines above](https://github.com/mantisbt/mantisbt/pull/1614/files#diff-aac9367df2352143e7002c2d9fbb1dd3R232).

Reply:
I'm aware of that, I was wondering if it would be helpful where logs from concurrent requests are interleaved.  Not a big deal though.

Reply:
Ah OK I see what you mean now. I didn't realize you were talking about concurrent requests. I'll update the log message.